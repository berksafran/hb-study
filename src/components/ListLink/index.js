import React from "react";
import { Button } from "react-bootstrap";
import {
  LinkListContainer,
  LinkDetail,
  LinkDetailNameAndURL,
  LinkDetailVoteButtons
} from "./style";
import { MdRemoveCircle } from "react-icons/md";
import { FaArrowUp, FaArrowDown } from "react-icons/fa";
import PropTypes from 'prop-types';

function ListLink(props) {
  const {linkName, linkURL, linkPoints, linkID} = props;
  return (
    <LinkListContainer id={linkName}>
      <Button
        variant="outline-secondary"
        size="large"
      >
        <p className="points">{linkPoints}</p>
        <p className="pointsText">POINTS</p>
      </Button>
      <LinkDetail>
        <LinkDetailNameAndURL>
          <p id="linkName">{linkName}</p>
          <small id="linkURL">({linkURL})</small>
        </LinkDetailNameAndURL>
        <LinkDetailVoteButtons>
          <div id="upVote" onClick={() => props.upVote(linkID)}>
            <FaArrowUp />
            <p>Up Vote</p>
          </div>
          <div id="downVote" onClick={() => props.downVote(linkID)}>
            <FaArrowDown />
            <p>Down Vote</p>
          </div>
        </LinkDetailVoteButtons>
      </LinkDetail>
      <div className="removeItem" onClick={() => props.removeLink(linkID)}>
        <MdRemoveCircle className="removeItemIcon" />
      </div>
    </LinkListContainer>
  );
}

ListLink.propTypes = {
  linkName: PropTypes.string.isRequired,
  linkURL: PropTypes.string.isRequired,
  linkPoints: PropTypes.number.isRequired,
  linkID: PropTypes.number.isRequired,
  upVote: PropTypes.func.isRequired,
  downVote: PropTypes.func.isRequired,
  removeLink: PropTypes.func.isRequired,
};

export default ListLink;


