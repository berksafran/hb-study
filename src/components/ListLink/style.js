import styled from "styled-components";

const LinkListContainer = styled.div`
  /* max-width: 85%; */
  /* width: 500px; bunu değiştir!! */
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  padding: 5px;
  position: relative;
  border-width: 1px;
  border-color: white;
  border-style: solid;
  border-radius: 7px;
  margin-bottom: 1.5em;

  p {
    margin: 0;
    padding: 0;
  }

  button,
  button:hover {
    font-size: 2rem;
    font-weight: bold;
    padding: 5px 20px 15px 20px;
    margin: 5px 5px 5px 5px;
    color: black;
    background-color: #ececec;
    border-color: gray;
    line-height: 1;
    cursor: default;

    .points {
      font-size: 2.5rem;
      font-weight: bold;
    }

    .pointsText {
      font-size: 1rem;
      font-weight: 300;
      padding-top: 0.5rem;
    }
  }

  .removeItem {
    display: none;
    font-size: 1.5em;
    position: absolute;
    right: -12px;
    top: -12px;
    padding: 0;
    margin: 0;

    .removeItemIcon {
      background-color: white;
      color: #dc3545;
      border-radius: 100px;
    }
  }

  :hover {
    background-color: #f9f9f9;
    border-width: 1px;
    border-color: lightgray;
    border-style: solid;
    border-radius: 7px;
    position: relative;

    .removeItem {
      display: flex;

      :hover {
        cursor: pointer;
      }
    }
  }
`;

const LinkDetail = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  /* height: 100%; */
  height: 6em;
  justify-content: space-around;
  padding-left: 1em;
  `;

const LinkDetailNameAndURL = styled.div`
  p {
    font-size: 1.2em;
    font-weight: bold;
    color: black;
    line-height: 1;
    padding-right: 1em;
  }
  
  small {
    color: #7f7f7f;
    letter-spacing: 0.02em;
  }
  `;

const LinkDetailVoteButtons = styled.div`
  display: flex;
  flex-direction: row;
  color: #7f7f7f;
  font-size: 0.7em;
  /* margin-top: 1em; */


  p {
    font-weight: bold;
    padding-left: 0.2em;

    :hover {
      text-decoration: underline;
      cursor: pointer;
    }
  }

  div {
    display: flex;
    flex-direction: row;
    align-items: center;
    margin-right: 6em;
  }
`;

export {
  LinkListContainer,
  LinkDetail,
  LinkDetailNameAndURL,
  LinkDetailVoteButtons
};
