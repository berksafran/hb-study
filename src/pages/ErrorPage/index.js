import React from "react";
import { Button } from "react-bootstrap";
import { NavLink } from "react-router-dom";

function ErrorPage() {
  return (
    <div className="d-flex flex-column justify-content-start">
      <h1 className="mt-5 ml-5">Something went wrong!</h1>
      <strong>
        <NavLink to="/">
          <Button variant="outline-secondary" className="m-5">
            Return to home page
          </Button>
        </NavLink>
      </strong>
    </div>
  );
}

export default ErrorPage;
