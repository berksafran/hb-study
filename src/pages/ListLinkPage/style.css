.pagination > li > a
{
    color: black;
    font-weight: bold;
}

.pagination > li {
    margin-left: 5px;
}

.pagination > li:hover {
    background-color: black;
}

.pagination > .page-item.active .page-link {
    z-index: 3;
    color: #fff;
    background-color: #6c757d;
    border-color: #6c757d;
    font-weight: bold;
}
.pagination > .page-link:hover {
    z-index: 2;
    color: black;
    text-decoration: none;
    background-color: #e9ecef;
    border-color: #dee2e6;
}

.dropdown-item.active, .dropdown-item:active  {
    background-color: #6c757d !important;
}

.dropdown-menu {
    z-index: 5000 !important;
}
