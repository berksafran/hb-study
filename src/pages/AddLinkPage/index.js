import React, { useState } from "react";
import { Row, Form, Button } from "react-bootstrap";
import { Link } from "react-router-dom";
import { FaArrowLeft } from "react-icons/fa";
import { AddLinkPageContainer } from "./style";
import _ from "lodash";
import ToastWarn from "../../components/ToastWarn";

function AddLinkPage() {
  const [validated, setValidated] = useState(false);
  const [linkName, setLinkName] = useState();
  const [linkURL, setLinkURL] = useState();
  const [showToast, setShowToast] = useState({
    status: false,
    name: ''
  });

  // This variable is for using Ref and clear all inputs of add link form.
  let addLinkForm = '';

  const handleSubmit = event => {
    const form = event.currentTarget;

    // Check input forms
    if (form.checkValidity() === false) {
      event.preventDefault();
      event.stopPropagation();
      setValidated(true);
      event.preventDefault();
    } else {
      setValidated(true);
      event.preventDefault();

      // Creating new link 
      var currentTime = new Date().getTime()
      const linksArray = JSON.parse(localStorage.getItem("linksArray"));
      const newLinkItem = {
        id: currentTime,
        linkName: linkName,
        linkURL: linkURL,
        modifiedDate: currentTime,
        voteCounter: 0,
        points: 0
      };

      // Is there any data in Localstorage? If yes, add new item to array.
      if (_.isArray(linksArray)) {
        let newLinksArray = [...linksArray, newLinkItem];
        localStorage.setItem("linksArray", JSON.stringify(newLinksArray));
      } else {
        // If not, create new array.
        let newLinksArray = [newLinkItem];
        localStorage.setItem("linksArray", JSON.stringify(newLinksArray));
      }
      
      setShowToast({
        status: true,
        name: linkName
      });

      addLinkForm.reset();
      setValidated(false);
    }
  };

   // Open or Close Toast
   function handleToast() {
    setShowToast({
      ...showToast,
      status: !showToast,
      name: linkName
    });
  }

  return (
    <AddLinkPageContainer>
    <ToastWarn showToast={showToast} handleToast={() => handleToast()} toastType="add" />
      <Row className="returnHome" id="returnHome">
        <FaArrowLeft className="icon" />
        <Link to="/">Return to List</Link>
      </Row>
      <Row className="mt-0">
        <h2 className="title">Add New Link</h2>
      </Row>
      <Row className="mt-0">
        <Form
          noValidate
          validated={validated}
          onSubmit={handleSubmit}
          className="addLinkForm"
          ref={ref => addLinkForm = ref}
        >
          <Form.Label className="inputLabel">Link Name:</Form.Label>
          <Form.Row>
            <Form.Group md="4" controlId="linkName" className="inputText">
              <Form.Control
                required
                onChange={e => setLinkName(e.target.value)}
                type="text"
                placeholder="e.g. Alphabet"
              />
            </Form.Group>
          </Form.Row>
          <Form.Label className="inputLabel">Link URL:</Form.Label>
          <Form.Row>
            <Form.Group md="4" controlId="linkURL" className="inputText">
              <Form.Control
                required
                type="url"
                placeholder="e.g. http://abc.xyz"
                onChange={e => setLinkURL(e.target.value)}
                ref={ref => {ref = linkURL}}
              />
            </Form.Group>
          </Form.Row>
          <Form.Row className="d-flex justify-content-end">
            <Button id="submitButton" className="btn btn-secondary submitButton" type="submit">
              ADD
            </Button>
          </Form.Row>
        </Form>
      </Row>
    </AddLinkPageContainer>
  );
}

export default AddLinkPage;
