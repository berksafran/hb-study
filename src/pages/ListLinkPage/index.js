import React, { useState } from "react";
import { Row, DropdownButton, Dropdown, Pagination } from "react-bootstrap";
import AddLink from "../../components/AddLink";
import Divider from "../../components/Divider";
import ListLink from "../../components/ListLink";
import RemoveModal from "../../components/RemoveModal";
import ToastWarn from "../../components/ToastWarn";
import _ from "lodash";
import "./style.css";

function ListLinkPage() {
  const [linksArray, setLinksArray] = useState(
    JSON.parse(localStorage.getItem("linksArray"))
  );
  const [showModal, setShowModal] = useState({
    status: false,
    name: "",
    id: ""
  });
  const [showToast, setShowToast] = useState({
    status: false,
    name: ""
  });
  const [orderType, setOrderType] = useState("mostVoted");
  const [selectedPage, setSelectedPage] = useState(0);

  // Up vote for specific link
  function upVote(linkID) {
    const targetItemIndex = _.findIndex(linksArray, o => o.id === linkID);
    let newArray = [...linksArray];
    ++newArray[targetItemIndex].voteCounter;
    ++newArray[targetItemIndex].points;
    newArray[targetItemIndex].modifiedDate = new Date().getTime();
    setLinksArray(newArray);
    localStorage.setItem("linksArray", JSON.stringify(newArray));
  }

  // Down vote for specific link
  function downVote(linkID) {
    const targetItemIndex = _.findIndex(linksArray, o => o.id === linkID);
    let newArray = [...linksArray];
    ++newArray[targetItemIndex].voteCounter;
    --newArray[targetItemIndex].points;
    newArray[targetItemIndex].modifiedDate = new Date().getTime();
    setLinksArray(newArray);
    localStorage.setItem("linksArray", JSON.stringify(newArray));
  }

  // Remove link
  function removeLink(linkID) {
    let newArray = _.filter(linksArray, o => o.id !== linkID);
    let removedItemName = _.filter(linksArray, o => o.id === linkID);
    setLinksArray([...newArray]);
    localStorage.setItem("linksArray", JSON.stringify(newArray));
    handleModal();
    setShowToast({
      status: true,
      name: removedItemName[0].linkName
    });
  }

  // Open or Close Modal
  function handleModal() {
    setShowModal({
      status: !showModal,
      name: "",
      id: ""
    });
  }

  // Open or Close Toast
  function handleToast() {
    setShowToast({
      ...showToast,
      status: !showToast
    });
  }

  // Order by Vote Counts
  const newLinksArray = () => {
    function orderArray() {
      // if (orderType === "mostVoted") {
      //   return _.orderBy(
      //     linksArray,
      //     ["voteCounter", "modifiedDate"],
      //     ["desc", "desc"]
      //   );
      // } else if (orderType === "mostVoted") {
      //   return _.orderBy(
      //     linksArray,
      //     ["voteCounter", "modifiedDate"],
      //     ["asc", "asc"]
      //   );
      // }

      switch (orderType) {
        case "mostVoted":
          return _.orderBy(
            linksArray,
            ["voteCounter", "modifiedDate"],
            ["desc", "desc"]
          );
        case "lessVoted":
          return _.orderBy(
            linksArray,
            ["voteCounter", "modifiedDate"],
            ["asc", "desc"]
          );
        case "bestRated":
          return _.orderBy(
            linksArray,
            ["points", "modifiedDate"],
            ["desc", "desc"]
          );
        case "worstRated":
          return _.orderBy(
            linksArray,
            ["points", "modifiedDate"],
            ["asc", "desc"]
          );
        default:
          return null;
      }
    }
    return _.chunk(orderArray(), 5);
  };

  // Pagination
  let items = [];
  function pagination() {
    for (let number = 1; number <= newLinksArray().length; number++) {
      items.push(
        <div key={number} className="pagination">
          <Pagination.Item
            key={number}
            active={number === selectedPage + 1}
            onClick={() => setSelectedPage(number - 1)}
            className="pagination"
          >
            {number}
          </Pagination.Item>
        </div>
      );
    }
  }

  pagination();

  return (
    <div>
      <Row className="d-flex justify-content-center m-4">
        <AddLink />
      </Row>
      <Divider />
      <Row className="d-flex justify-content-center m-4 p-2">
        <DropdownButton
          id="orderbyButton"
          title="Order by"
          variant="secondary"
          block="true"
        >
          <Dropdown.Item onClick={() => setOrderType("mostVoted")} active={orderType === 'mostVoted' ? true : false}>
            Most Voted
          </Dropdown.Item>
          <Dropdown.Item onClick={() => setOrderType("lessVoted")} active={orderType === 'lessVoted' ? true : false}>
            Less Voted
          </Dropdown.Item>
          <Dropdown.Divider />
          <Dropdown.Item onClick={() => setOrderType("bestRated")} active={orderType === 'bestRated' ? true : false}>
            Best Rated
          </Dropdown.Item>
          <Dropdown.Item onClick={() => setOrderType("worstRated")} active={orderType === 'worstRated' ? true : false}>
            Worst Rated
          </Dropdown.Item>
        </DropdownButton>
      </Row>
      <Row className="d-flex justify-content-center m-4">
        <Row className="d-flex flex-column">
          {_.size(linksArray) === 0 ? (
            <div>
              <br />
              <h3>Do you want to add first link here?</h3>
            </div>
          ) : (
            _.map(newLinksArray()[selectedPage], (link, index) => (
              <ListLink
                key={index}
                linkID={link.id}
                linkName={link.linkName}
                linkURL={link.linkURL}
                linkPoints={link.points}
                linkVoteCounter={link.voteCounter}
                upVote={linkID => upVote(linkID)}
                downVote={linkID => downVote(linkID)}
                removeLink={() => {
                  setShowModal({
                    status: true,
                    name: link.linkName,
                    id: link.id
                  });
                }}
              />
            ))
          )}
        </Row>
      </Row>
      <Row className="d-flex justify-content-center m-4">
        <Pagination>{items}</Pagination>
      </Row>
      <RemoveModal
        showModal={showModal}
        handleModal={handleModal}
        removeLink={linkID => removeLink(linkID)}
      />
      <ToastWarn
        showToast={showToast}
        handleToast={() => handleToast()}
        toastType="remove"
      />
    </div>
  );
}

export default ListLinkPage;
