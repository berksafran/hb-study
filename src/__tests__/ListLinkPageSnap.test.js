import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import ListLink from "../components/ListLink/index";
import { upVote, downVote, removeLink } from "../pages/ListLinkPage";

describe("ListLinkPage snapshot", () => {
  test("Header component Snapshot", () => {
    const header_tree = renderer
      .create(
        <ListLink
          linkID={234234112}
          linkName="Amazon"
          linkURL="http://www.amazon.com"
          linkPoints={9999}
          upVote={() => upVote()}
          downVote={() => downVote()}
          removeLink={() => removeLink()}
        />
      )
      .toJSON();
    expect(header_tree).toMatchSnapshot();
  });

  test("ListLink component snapshot", () => {
    const wrapper = shallow(
      <ListLink
        linkID={12312312}
        linkName="Stackoverflow"
        linkURL="http://www.stackoverflow.com"
        linkPoints={232}
        upVote={() => upVote()}
        downVote={() => downVote()}
        removeLink={() => removeLink()}
      />
    );

    expect(wrapper.find(".points")).toHaveLength(1);
  });
});
