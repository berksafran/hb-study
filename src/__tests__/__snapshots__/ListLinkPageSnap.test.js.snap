// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListLinkPage snapshot Header component Snapshot 1`] = `
<div
  className="sc-bdVaJa heZbqI"
  id="Amazon"
>
  <button
    className="btn btn-outline-secondary btn-large"
    disabled={false}
    type="button"
  >
    <p
      className="points"
    >
      9999
    </p>
    <p
      className="pointsText"
    >
      POINTS
    </p>
  </button>
  <div
    className="sc-bwzfXH esojNt"
  >
    <div
      className="sc-htpNat WoJPU"
    >
      <p
        id="linkName"
      >
        Amazon
      </p>
      <small
        id="linkURL"
      >
        (
        http://www.amazon.com
        )
      </small>
    </div>
    <div
      className="sc-bxivhb kALBOD"
    >
      <div
        id="upVote"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
          />
        </svg>
        <p>
          Up Vote
        </p>
      </div>
      <div
        id="downVote"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
          />
        </svg>
        <p>
          Down Vote
        </p>
      </div>
    </div>
  </div>
  <div
    className="removeItem"
    onClick={[Function]}
  >
    <svg
      className="removeItemIcon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11H7v-2h10v2z"
      />
    </svg>
  </div>
</div>
`;
